rust/actyx/api/src/auth/mod.rs

```diff
@@ -93,10 +93,11 @@ fn validate_manifest(
     ax_public_key: &PublicKey,
     licensing: &Licensing,
 ) -> Result<(AppMode, AppId, String), ApiError> {
-    match manifest {
-        AppManifest::Signed(x) => validate_signed_manifest(x, ax_public_key, licensing)
-            .map(|_| (AppMode::Signed, x.get_app_id(), x.version.clone())),
-        AppManifest::Trial(x) => Ok((AppMode::Trial, x.get_app_id(), x.version.clone())),
+    if manifest.is_signed() {
+        validate_signed_manifest(manifest, ax_public_key, licensing)
+            .map(|_| (AppMode::Signed, manifest.app_id(), manifest.version().to_owned()))
+    } else {
+        Ok((AppMode::Trial, manifest.app_id(), manifest.version().to_owned()))
```

Comment:
I know the PR disallows creating "invalid" trials in the Rust code, but since this is part of a REST endpoint, we might be receiving requests from other SDKs, etc

Something among the lines of:
```suggestion
    } else if manifest.app_id().starts_with("com.example.") {
        Ok((AppMode::Trial, manifest.app_id(), manifest.version().to_owned()))
    } else {
        Err(ApiError::InvalidManifest { msg: "Trial application IDs are required to start with "com.example." }
    }
```

Reply:
This is exactly the reason for the wrapper type: AppManifestIo is the serialisation format that comes off the wire, which validates only the basic data types, while AppManifest firmly establishes the trial manifest validation rule (see `impl TryFrom`). I searched the serde github repo for confirmation, this is the intended implementation pattern of deserialisation-is-validation (i.e. `#[serde(validate = ...)]` were all rejected with this reason).